loom/engine/bindings/loom/lmSQLite.cpp

```diff
@@ -0,0 +1,467 @@
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ===========================================================================
+ */
+
+#include "platform/CCFileUtils.h"
+#include "loom/script/loomscript.h"
+#include "loom/script/runtime/lsRuntime.h"
+#include "loom/common/core/log.h"
+#include "loom/script/native/lsNativeDelegate.h"
+#include "loom/vendor/sqlite3/sqlite3.h"
+#include "loom/common/utils/utByteArray.h"
+#include "loom/common/utils/utString.h"
+ #include "loom/common/platform/platformThread.h"
+
+
+lmDefineLogGroup(gSQLiteGroup, "loom.sqlite", 1, LoomLogInfo);
+
+using namespace LS;
+
+
+//forward declaration of Statement
+class Statement;
+
+
+//SQLite Connection binding for Loomscript
+class Connection
+{
+protected:
+    utString databaseName;
+    utString databaseFullPath;
+    sqlite3 *dbHandle;
+
+public:
+    LOOM_STATICDELEGATE(OnImportComplete);
+
+    static Connection *open(const char *database, const char *path, int flags);
+    static void backgroundImport(const char *database, const char *path, utByteArray *data);
+    static const char *getVersion();
+
+    Statement *prepare(const char *query);
+    const char* getDBName() { return databaseName.c_str(); }
+
+
+    int getErrorCode()
+    {
+        return sqlite3_errcode(dbHandle);
+    }
+  
+    const char* getErrorMessage()
+    {
+        return sqlite3_errmsg(dbHandle);
+    }
+      
+    int getlastInsertRowId()
+    {
+        //"NOTE: In SQLite the row ID is a 64-bit integer but for all practical 
+        //database sizes you can cast the 64 bit value to a 32-bit integer."
+        //
+        // - Some Guy on The Internet
+        //
+        sqlite3_int64 rowid64 = sqlite3_last_insert_rowid(dbHandle);
+
+        //safety check on the value of the row as SQLite allows 64bit ints and Loomscript 
+        //only supports 32bit (31 for signed)
+        if(rowid64 > (2^31))
+        {
+            lmLogError(gSQLiteGroup, "RowID found in getlastInsertRowId the SQLite database %s is larger than a 32 bit integer! The return value will not be as expected!", getDBName());
+        }
+        return (int)rowid64;
+    }
+
+    int close()
+    {
+        //close the database
+        int result = sqlite3_close_v2(dbHandle);
+        if(result != SQLITE_OK)
+        {
+            lmLogError(gSQLiteGroup, "Error closing the SQLite database: %s with message: %s", getDBName(), getErrorMessage());
+        }
+        return result;
+    }
+};
+
+
+
+//SQLite Statement binding for Loomscript
+class Statement
+{
+public:
+    LOOM_DELEGATE(OnStatementProgress);
+    LOOM_DELEGATE(OnStatementComplete);
+
+    MutexHandle stepAsyncDelegateMutex;
```

Comment:
Cool, so I can do away with the mutex then?
On 12 Feb 2015 10:53 PM, "Ben Garney" <notifications@github.com> wrote:

> In loom/engine/bindings/loom/lmSQLite.cpp
> <https://github.com/LoomSDK/LoomSDK/pull/261#discussion_r24619047>:
>
> > +            lmLogError(gSQLiteGroup, "Error closing the SQLite database: %s with message: %s", getDBName(), getErrorMessage());
> > +        }
> > +        return result;
> > +    }
> > +};
> > +
> > +
> > +
> > +//SQLite Statement binding for Loomscript
> > +class Statement
> > +{
> > +public:
> > +    LOOM_DELEGATE(OnStatementProgress);
> > +    LOOM_DELEGATE(OnStatementComplete);
> > +
> > +    MutexHandle stepAsyncDelegateMutex;
>
> Delegates are threadsafe now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/LoomSDK/LoomSDK/pull/261/files#r24619047>.
>